networkx/algorithms/similarity.py

```diff
@@ -1386,3 +1393,249 @@ def simrank_similarity_numpy(
+    similarity : dictionary
+        Dictionary of nodes to similarity scores (as floats). Note:
+        the self-similarity (i.e., ``v``) will not be included in
+        the returned dictionary.
+
+    Examples
+    --------
+
+        >>> import networkx as nx
+        >>> G = nx.star_graph(10)
+        >>> sim = nx.panther_similarity(G, 0)
+
+    References
+    ----------
+    .. [1] Zhang, J., Tang, J., Ma, C., Tong, H., Jing, Y., & Li, J.
+           Panther: Fast top-k similarity search on large networks.
+           In Proceedings of the ACM SIGKDD International Conference
+           on Knowledge Discovery and Data Mining (Vol. 2015-August, pp. 1445–1454).
+           Association for Computing Machinery. https://doi.org/10.1145/2783258.2783267.
+    """
+    import numpy as np
+
+    num_nodes = G.number_of_nodes()
+    if num_nodes < k:
+        warnings.warn(
+            f"Number of nodes is {num_nodes}, but requested k is {k}. "
+            "Setting k to number of nodes."
+        )
+        k = num_nodes
+    # According to [1], they empirically determined
+    # a good value for ``eps`` to be sqrt( 1 / |E| )
+    if eps is None:
+        eps = np.sqrt(1.0 / G.number_of_edges())
+
+    inv_node_map = {name: index for index, name in enumerate(G.nodes)}
+    node_map = np.array(G)
+
+    # Calculate the sample size ``R`` for how many paths
+    # torandomly generate
+    t_choose_2 = n_choose_k(path_length, 2)
+    sample_size = int((c / eps ** 2) * (np.log2(t_choose_2) + 1 + np.log(1 / delta)))
+    index_map = {}
+    paths = generate_random_paths(
+        G,
+        sample_size,
+        path_length=path_length,
+        index_map=index_map
+    )
+    S = np.zeros(G.number_of_nodes())
+
+    # Calculate the path similarities
+    for path_index, path in enumerate(paths):
+        path_set = set(path)
+
+        # Comparing ``v`` with ``node`` (v_j)
+        for node in path_set:
+            # Don't compare with self
+            if v == node:
+                continue
+
+            node_index = inv_node_map[node]
+
+            # Only sum if they are share the same path,
+            # i.e., ``v`` is also on the same path
+            if path_index in index_map[v]:
+                S[node_index] += 1 / sample_size
+
+    # Retrieve top ``k`` similar
+    # Note: the below performed anywhere from 4-10x faster
+    # (depending on input sizes) vs the equivalent ``np.argsort(S)[::-1]``
+    top_k_unsorted = np.argpartition(S, -k)[-k:]
+    top_k_sorted = top_k_unsorted[np.argsort(S[top_k_unsorted])][::-1]
+
+    # Add back the similarity scores
+    top_k_sorted_names = map(lambda n: node_map[n], top_k_sorted)
+    top_k_with_val = dict(zip(top_k_sorted_names, S[top_k_sorted]))
+
+    # Remove the self-similarity
+    top_k_with_val.pop(v, None)
+    return top_k_with_val
+
+
+def generate_random_paths(G, sample_size, path_length=5, index_map=None):
+    """Randomly generate ``sample_size`` paths of length ``path_length``.
```

Comment:
Would ```generate_random_paths``` be better as a generator yielding paths?

Reply:
That makes sense to me. Addressed in 49302de.